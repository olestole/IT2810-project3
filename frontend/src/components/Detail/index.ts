import DetailView from './DetailView';
import ProductReview from './ProductReview/ProductReview';

export { DetailView, ProductReview };
