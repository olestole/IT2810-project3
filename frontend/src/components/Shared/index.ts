import Search from './Search';
import Header from './Header';
import Logo from './Logo';
import ReviewItem from './Review/ReviewItem';
import ReviewList from './Review/ReviewList';

export { Search, Header, Logo, ReviewItem, ReviewList };
