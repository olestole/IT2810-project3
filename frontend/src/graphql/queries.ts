import { gql } from '@apollo/client';

const GET_REVIEWS = gql`
  query GetReviewsQuery($reviewsVarenummer: String!) {
    reviews(varenummer: $reviewsVarenummer) {
      userEmail
      varenummer
      title
      description
      rating
    }
  }
`;

const GET_PERSONAL_REVIEWS = gql`
  query GetPersonalReviewsQuery($personalReviewUserEmail: String!) {
    personalReviews(userEmail: $personalReviewUserEmail) {
      userEmail
      varenummer
      title
      description
      rating
    }
  }
`;

const GET_SINGLE_PRODUCT = gql`
  query SingleProductQuery($number: String!) {
    singleProduct(productNumber: $number) {
      Varenummer
      Varenavn
      Volum
      Pris
      Varetype
      Farge
      Lukt
      Smak
      Land
      Produsent
    }
  }
`;

const GET_START_PRODUCTS = gql`
  query StartProductsQuery($index: Int!) {
    startProducts(startIndex: $index) {
      Varenavn
      Varenummer
      Varetype
      Varenummer
      Produsent
      Volum
      Pris
    }
  }
`;

const SEARCH_PRODUCTS = gql`
  query SearchProductsQuery($matchedString: String!, $searchIndex: Int!) {
    searchProducts(searchSequence: $matchedString, index: $searchIndex) {
      Varenavn
      Varetype
      Varenummer
      Produsent
      Volum
      Pris
    }
  }
`;

const FILTER_PRODUCTS = gql`
  query FilterProductsQuery(
    $typer: [String]!
    $prisgt: Float!
    $prisls: Float!
    $volumgt: Float!
    $volumls: Float!
    $filterIndex: Int!
  ) {
    filterProducts(
      varetyper: $typer
      prisgt: $prisgt
      prisls: $prisls
      volumgt: $volumgt
      volumls: $volumls
      index: $filterIndex
    ) {
      Varenavn
      Varetype
      Varenummer
      Produsent
      Volum
      Pris
    }
  }
`;

export { GET_START_PRODUCTS, SEARCH_PRODUCTS, GET_REVIEWS, GET_PERSONAL_REVIEWS, GET_SINGLE_PRODUCT, FILTER_PRODUCTS };
// export { GET_REVIEWS, GET_PERSONAL_REVIEWS, GET_SINGLE_PRODUCT };
