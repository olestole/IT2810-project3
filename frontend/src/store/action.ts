import { Product } from 'types/globalTypes';

export const increment = () => {
  return {
    type: 'INCREMENT',
  } as const;
};

export const decrement = () => {
  return {
    type: 'DECREMENT',
  } as const;
};

export const setSearchText = (text: string) => {
  return {
    type: 'SET_SEARCH_TEXT',
    payload: text,
  } as const;
};

export const setCurrentProduct = (product: Product | null) => {
  return {
    type: 'SET_CURRENT_PRODUCT',
    payload: product,
  } as const;
};

export const setModalOpen = (value: boolean) => {
  return {
    type: 'SET_MODAL_OPEN',
    payload: value,
  } as const;
};
