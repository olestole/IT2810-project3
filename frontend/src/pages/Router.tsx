import React from 'react';
import { Switch, Route } from 'react-router-dom';
import { Callback } from './Callback';
import Detail from './Detail';
import Login from './Login';
import Overview from './Overview/Overview';

const Router = () => (
  <div className="layout">
    <Switch>
      <Route path="/login" component={Login} />
      <Route path="/callback" component={Callback} />
      <Route path="/:id" component={Detail} />
      <Route exact path="/" component={Overview} />
    </Switch>
  </div>
);

export default Router;
