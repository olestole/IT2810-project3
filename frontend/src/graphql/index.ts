import { GET_START_PRODUCTS, SEARCH_PRODUCTS } from './queries';
import { ADD_REVIEW } from './mutations';
export { GET_START_PRODUCTS, SEARCH_PRODUCTS, ADD_REVIEW };
