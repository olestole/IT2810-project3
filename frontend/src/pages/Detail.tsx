import { gql, useLazyQuery, useQuery } from '@apollo/client';
import React from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { useLocation } from 'react-router-dom';
import { decrement, increment } from 'store/action';
import { AppState } from 'store/types';
import { makeQuery } from 'utils/client';

const GET_WHITE_WINES = gql`
  query Query {
    whiteWines {
      Varenavn
      Land
      Distrikt
    }
  }
`;

const GET_SINGLE_PRODUCT = gql`
  query Query($number: String!) {
    singleProduct(productNumber: $number) {
      Varenavn
      Varenummer
    }
  }
`;

const Detail = () => {
  const location = useLocation();
  const { data, loading, error } = useQuery(GET_SINGLE_PRODUCT, { variables: { number: location.pathname.substr(1) } });

  // REDUX BELOW
  const dispatch = useDispatch();
  const count: number = useSelector((state: AppState) => state.count);

  const dispatchIncrement = (e: any) => {
    e.preventDefault();
    dispatch(increment());
  };
  const dispatchDecrement = (e: any) => {
    e.preventDefault();
    dispatch(decrement());
  };
  // REDUX ABOVE

  //useLazyQuery return a function which can be used to trigger the query manually and we should use this for dynamic loading
  if (loading) return <p>Loading ...</p>;

  if (data && data.singleProduct) {
    console.log(data.singleProduct);
  }

  return (
    <div>
      <h1>Detail</h1>
      <h1>Count: {count}</h1>
      <button onClick={dispatchIncrement}>INCREMENT</button>
      <button onClick={dispatchDecrement}>DECREMENT</button>
      <h3>{data.singleProduct.Varenavn}</h3>
      <h3>{data.singleProduct.Varetype}</h3>
    </div>
  );
};

export default Detail;

