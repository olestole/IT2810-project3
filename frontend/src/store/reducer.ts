import { createStore, combineReducers, Store } from 'redux';
import { decrement, increment, setSearchText, filter, filterVolumAndPrice, setFilterMode } from './action';
import { AppState, FilterOptions } from './types';

type Actions = ReturnType<typeof increment> | ReturnType<typeof decrement> | ReturnType<typeof setFilterMode> | ReturnType<typeof setSearchText> | ReturnType<typeof filter> | ReturnType<typeof filterVolumAndPrice>;

const initialAppState: AppState = {
  count: 0,
  searchText: "",
  filterOptions: {
    filterMode: false,
    kategorier: {
      rodvin: false,
      hvitvin: false,
      musserende_vin: false,
      sterk_vin: false,
      annen_vin: false,
      ol: false,
      brennevin: false,
      alkoholfritt: false,
      annet: false,
    },
    minVolum: 0, 
    maxVolum: 10,
    minPrice: 0, 
    maxPrice: 2000000,
  },
};

const rootReducer = (state: AppState = initialAppState, action: Actions) => {
  switch (action.type) {
    case 'INCREMENT':
      return {
        ...state,
        count: state.count + 1,
      };
    case 'DECREMENT':
      return {
        ...state,
        count: state.count - 1,
      };
    case 'FILTER_MODE':
      return {
        ...state,
        filterOptions: {
          ...state.filterOptions,
          filterMode: action.payload,
        }
      };
    case 'SET_SEARCH_TEXT':
      return {
        ...state,
        searchText: action.payload
      };
    case 'FILTER':
      return {
        ...state,
        filterOptions: {
            ...state.filterOptions,
            kategorier: {
              ...state.filterOptions.kategorier,
              [action.payload.field]: action.payload.value  
          }
        }
      };
    case 'FILTER_RANGE':
      return {
        ...state,
        filterOptions: {
            ...state.filterOptions,
            [action.payload.field]: action.payload.value  
        }
      }
    default:
      neverReached(action);
  }
  return state;
};

const neverReached = (never: never) => {};

//Utility-funksjon for Ã¥ kombinere flere reducere
// const rootReducer = combineReducers<AppState>({
//   count: countReducer,
// });

function configureStore(): Store<AppState> {
  return createStore(
    rootReducer,
    (window as any).__REDUX_DEVTOOLS_EXTENSION__ && (window as any).__REDUX_DEVTOOLS_EXTENSION__(),
  );
}

//Oppretter en store
export const store = configureStore();
