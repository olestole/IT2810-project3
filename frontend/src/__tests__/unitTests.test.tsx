import React from 'react';
import {render, fireEvent, RenderResult, cleanup, screen } from '@testing-library/react'
import { Provider } from 'react-redux';
import { store } from 'store/reducer';
import { Logo, Header, Search } from 'components/Shared'
import { BrowserRouter} from 'react-router-dom';
import Router from 'pages/Router';
import { CustomToastContainer } from 'components/Shared/FeedbackToast';
import ApolloWrapper from 'utils/ApolloWrapper';
import { createMemoryHistory } from 'history';
import * as actions from 'store/action';
import { MockedProvider } from '@apollo/client/testing'
import Detail from 'pages/Detail'
import { GET_SINGLE_PRODUCT, GET_REVIEWS } from 'graphql/queries';

let documentBody: RenderResult;

afterEach(cleanup)

//testing logo component with user event (homeclick)
describe('logo', () => {
  beforeEach(() => {
    documentBody = render(<ApolloWrapper>
      <BrowserRouter>
        <Provider store={store}>
          <CustomToastContainer />
          <Logo />
          <Router />
        </Provider>
      </BrowserRouter>
    </ApolloWrapper>
  )})

  it('renders correctly', () => {
        expect(documentBody.getByRole('title')).toBeTruthy()
        expect(documentBody.getByRole('image')).toBeTruthy()
  })

  it('shows title of page', () => {
    expect(documentBody.getByText("WineEncyclopedia")).toBeInTheDocument()
  })

  it('handle home click', () => {
    const { getByRole } = documentBody
    const history = createMemoryHistory()

    fireEvent.click(getByRole('homeButton'));

    expect(history.location.pathname).toBe('/')
  })

  it('matches snapshot', () => {
    expect(documentBody).toMatchSnapshot()
  })

});


//Testing search component with user event (search input and function)
describe('search', () => {
  beforeEach(() => {
    documentBody = render(<ApolloWrapper>
      <BrowserRouter>
        <Provider store={store}>
          <CustomToastContainer />
          <Search />
        </Provider>
      </BrowserRouter>
    </ApolloWrapper>
  )})


  it('renders correctly', () => {
    expect(documentBody.getByLabelText('searchIcon')).toBeTruthy()
    expect(documentBody.getByPlaceholderText('Produkt…')).toBeTruthy()
  })

  it('shows chip when search', () => {
    fireEvent.change(documentBody.getByPlaceholderText('Produkt…'), { target: { value: "whiskey" } })
    fireEvent.keyDown(documentBody.getByPlaceholderText('Produkt…'), { key: 'Enter', keyCode: 13, charCode: 13 })

    expect(documentBody.getByRole('chip')).toHaveTextContent('whiskey');
  })
});


//Testing actions from store
describe('actions', () => {
  it('should set search string', () => {
    const text = 'whiskey'
    const expectedAction = {
      type: 'SET_SEARCH_TEXT',
      payload: text
    }
    expect(actions.setSearchText(text)).toEqual(expectedAction)
  })

  it('should set filtermode', () => {
    const boolean = true
    const expectedAction = {
      type: 'FILTER_MODE',
      payload: boolean
    }
    expect(actions.setFilterMode(boolean)).toEqual(expectedAction)
  })
})

//Mocks for mocking the two graphql queries GET_SINGLE_PRODUCT and GET_REVIEWS
const mocks= [{
  request: {
    query: GET_SINGLE_PRODUCT,
    variables: {
      number: location.pathname.substr(1)
    }
  },
  result: {
    data: {
      singleProduct: {
        Varenummer: "232101",
        Varenavn: "Dom Perignon Brut 2010",
        Volum: "0.75",
        Pris: "1649.9",
        Varetype: "Champagne, brut",
        Farge: "Middels dyp grønngul.",
        Lukt: "Fokusert aroma av eple og sitrus, litt blomst og steinfrukt over innslag av kjeks og mineraler.",
        Smak: "Litt utviklet, kremet og frisk, preg av eple, sitrus, mineraler, nøtt og kjeks. Lang ettersmak.",
        Land: "Frankrike",
        Produsent: "Moët & Chandon",
      }
    }
  },

},
{
  request: {
    query: GET_REVIEWS,
    variables: {
      reviewsVarenummer: "232101"
    },
    fetchPolicy: 'network-only',
  }, 
  result: {
    data: {
      reviews: [
        {
          userEmail: "oleastole@gmail.com",
          varenummer: "232101",
          title: "Nydelig guttastemning",
          description: "Drakk denne på fest og ble damenes midtpunkt på en, to, tre! Anbefales\n\n-Henrik Gundersen (23, Holmenkollen)",
          rating: 5
        }
      ]
    }
  }
}]

//Test for testing graphql queries with the mocks from above
describe('graphql queries', () => {
  it('get single product with user reviews', async () => {
    const { findByText, getByRole } = render(
      <BrowserRouter>
        <Provider store={store}>
          <MockedProvider mocks={mocks} addTypename={false}>
          <Detail />
         </MockedProvider>
        </Provider>
      </BrowserRouter>
   )
    expect(getByRole("loading")).toBeInTheDocument();
    
    const itemName = await findByText("Dom Perignon Brut 2010");
    const title = await findByText("Nydelig guttastemning");
    expect(itemName).toBeInTheDocument();
    expect(title).toBeInTheDocument();
  })
})



//Snapshot test for the app
describe('app', () => {
  it('matches snapshot', () => {
    const { asFragment } = render(<ApolloWrapper>
      <BrowserRouter>
        <Provider store={store}>
          <CustomToastContainer />
          <Header />
          <Router />
        </Provider>
      </BrowserRouter>
    </ApolloWrapper>)

    expect(asFragment).toMatchSnapshot()
  })
});

