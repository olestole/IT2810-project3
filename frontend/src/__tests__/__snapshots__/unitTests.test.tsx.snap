// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`app matches snapshot 1`] = `[Function]`;

exports[`logo matches snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="Toastify"
      />
      <div
        class="logoContainer"
        role="homeButton"
      >
        <div
          class="logo"
          role="image"
        >
          <img
            src="appLogo.svg"
          />
        </div>
        <div
          class="title"
        >
          <h4
            class="MuiTypography-root makeStyles-title-43 MuiTypography-h4 MuiTypography-noWrap MuiTypography-alignRight"
            role="title"
          >
            WineEncyclopedia
          </h4>
        </div>
      </div>
      <div
        class="layout"
      >
        <div
          class="overview"
        >
          <div
            class="sidebar"
          >
            <nav
              class="MuiList-root makeStyles-appMenu-44"
            >
              <div
                aria-disabled="false"
                class="MuiButtonBase-root MuiListItem-root makeStyles-menuItem-46 MuiListItem-gutters MuiListItem-button"
                id="kategoriId"
                role="button"
                tabindex="0"
              >
                <div
                  class="MuiListItemIcon-root makeStyles-menuItemIcon-47"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M3 13h8V3H3v10zm0 8h8v-6H3v6zm10 0h8V11h-8v10zm0-18v6h8V3h-8z"
                    />
                  </svg>
                </div>
                <div
                  class="MuiListItemText-root"
                >
                  <span
                    class="MuiTypography-root MuiListItemText-primary MuiTypography-body1 MuiTypography-displayBlock"
                  >
                    Kategori
                  </span>
                </div>
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M16.59 8.59L12 13.17 7.41 8.59 6 10l6 6 6-6z"
                  />
                </svg>
                <span
                  class="MuiTouchRipple-root"
                />
              </div>
              <div
                aria-disabled="false"
                class="MuiButtonBase-root MuiListItem-root makeStyles-menuItem-46 MuiListItem-gutters MuiListItem-button"
                role="button"
                tabindex="0"
              >
                <div
                  class="MuiListItemIcon-root makeStyles-menuItemIcon-47"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M3 2l2.01 18.23C5.13 21.23 5.97 22 7 22h10c1.03 0 1.87-.77 1.99-1.77L21 2H3zm9 17c-1.66 0-3-1.34-3-3 0-2 3-5.4 3-5.4s3 3.4 3 5.4c0 1.66-1.34 3-3 3zm6.33-11H5.67l-.44-4h13.53l-.43 4z"
                    />
                  </svg>
                </div>
                <div
                  class="MuiListItemText-root"
                >
                  <span
                    class="MuiTypography-root MuiListItemText-primary MuiTypography-body1 MuiTypography-displayBlock"
                  >
                    Volum
                  </span>
                </div>
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M16.59 8.59L12 13.17 7.41 8.59 6 10l6 6 6-6z"
                  />
                </svg>
                <span
                  class="MuiTouchRipple-root"
                />
              </div>
              <div
                aria-disabled="false"
                class="MuiButtonBase-root MuiListItem-root makeStyles-menuItem-46 MuiListItem-gutters MuiListItem-button"
                role="button"
                tabindex="0"
              >
                <div
                  class="MuiListItemIcon-root makeStyles-menuItemIcon-47"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M11.8 10.9c-2.27-.59-3-1.2-3-2.15 0-1.09 1.01-1.85 2.7-1.85 1.78 0 2.44.85 2.5 2.1h2.21c-.07-1.72-1.12-3.3-3.21-3.81V3h-3v2.16c-1.94.42-3.5 1.68-3.5 3.61 0 2.31 1.91 3.46 4.7 4.13 2.5.6 3 1.48 3 2.41 0 .69-.49 1.79-2.7 1.79-2.06 0-2.87-.92-2.98-2.1h-2.2c.12 2.19 1.76 3.42 3.68 3.83V21h3v-2.15c1.95-.37 3.5-1.5 3.5-3.55 0-2.84-2.43-3.81-4.7-4.4z"
                    />
                  </svg>
                </div>
                <div
                  class="MuiListItemText-root"
                >
                  <span
                    class="MuiTypography-root MuiListItemText-primary MuiTypography-body1 MuiTypography-displayBlock"
                  >
                    Pris
                  </span>
                </div>
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M16.59 8.59L12 13.17 7.41 8.59 6 10l6 6 6-6z"
                  />
                </svg>
                <span
                  class="MuiTouchRipple-root"
                />
              </div>
            </nav>
          </div>
          <div
            class="loadingIndicator"
            role="loading"
          >
            <div
              class="makeStyles-root-56"
            >
              <div
                class="MuiCircularProgress-root MuiCircularProgress-colorPrimary MuiCircularProgress-indeterminate"
                role="progressbar"
                style="width: 6rem; height: 6rem;"
              >
                <svg
                  class="MuiCircularProgress-svg"
                  viewBox="22 22 44 44"
                >
                  <circle
                    class="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate"
                    cx="44"
                    cy="44"
                    fill="none"
                    r="20.2"
                    stroke-width="3.6"
                  />
                </svg>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="Toastify"
    />
    <div
      class="logoContainer"
      role="homeButton"
    >
      <div
        class="logo"
        role="image"
      >
        <img
          src="appLogo.svg"
        />
      </div>
      <div
        class="title"
      >
        <h4
          class="MuiTypography-root makeStyles-title-43 MuiTypography-h4 MuiTypography-noWrap MuiTypography-alignRight"
          role="title"
        >
          WineEncyclopedia
        </h4>
      </div>
    </div>
    <div
      class="layout"
    >
      <div
        class="overview"
      >
        <div
          class="sidebar"
        >
          <nav
            class="MuiList-root makeStyles-appMenu-44"
          >
            <div
              aria-disabled="false"
              class="MuiButtonBase-root MuiListItem-root makeStyles-menuItem-46 MuiListItem-gutters MuiListItem-button"
              id="kategoriId"
              role="button"
              tabindex="0"
            >
              <div
                class="MuiListItemIcon-root makeStyles-menuItemIcon-47"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M3 13h8V3H3v10zm0 8h8v-6H3v6zm10 0h8V11h-8v10zm0-18v6h8V3h-8z"
                  />
                </svg>
              </div>
              <div
                class="MuiListItemText-root"
              >
                <span
                  class="MuiTypography-root MuiListItemText-primary MuiTypography-body1 MuiTypography-displayBlock"
                >
                  Kategori
                </span>
              </div>
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M16.59 8.59L12 13.17 7.41 8.59 6 10l6 6 6-6z"
                />
              </svg>
              <span
                class="MuiTouchRipple-root"
              />
            </div>
            <div
              aria-disabled="false"
              class="MuiButtonBase-root MuiListItem-root makeStyles-menuItem-46 MuiListItem-gutters MuiListItem-button"
              role="button"
              tabindex="0"
            >
              <div
                class="MuiListItemIcon-root makeStyles-menuItemIcon-47"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M3 2l2.01 18.23C5.13 21.23 5.97 22 7 22h10c1.03 0 1.87-.77 1.99-1.77L21 2H3zm9 17c-1.66 0-3-1.34-3-3 0-2 3-5.4 3-5.4s3 3.4 3 5.4c0 1.66-1.34 3-3 3zm6.33-11H5.67l-.44-4h13.53l-.43 4z"
                  />
                </svg>
              </div>
              <div
                class="MuiListItemText-root"
              >
                <span
                  class="MuiTypography-root MuiListItemText-primary MuiTypography-body1 MuiTypography-displayBlock"
                >
                  Volum
                </span>
              </div>
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M16.59 8.59L12 13.17 7.41 8.59 6 10l6 6 6-6z"
                />
              </svg>
              <span
                class="MuiTouchRipple-root"
              />
            </div>
            <div
              aria-disabled="false"
              class="MuiButtonBase-root MuiListItem-root makeStyles-menuItem-46 MuiListItem-gutters MuiListItem-button"
              role="button"
              tabindex="0"
            >
              <div
                class="MuiListItemIcon-root makeStyles-menuItemIcon-47"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M11.8 10.9c-2.27-.59-3-1.2-3-2.15 0-1.09 1.01-1.85 2.7-1.85 1.78 0 2.44.85 2.5 2.1h2.21c-.07-1.72-1.12-3.3-3.21-3.81V3h-3v2.16c-1.94.42-3.5 1.68-3.5 3.61 0 2.31 1.91 3.46 4.7 4.13 2.5.6 3 1.48 3 2.41 0 .69-.49 1.79-2.7 1.79-2.06 0-2.87-.92-2.98-2.1h-2.2c.12 2.19 1.76 3.42 3.68 3.83V21h3v-2.15c1.95-.37 3.5-1.5 3.5-3.55 0-2.84-2.43-3.81-4.7-4.4z"
                  />
                </svg>
              </div>
              <div
                class="MuiListItemText-root"
              >
                <span
                  class="MuiTypography-root MuiListItemText-primary MuiTypography-body1 MuiTypography-displayBlock"
                >
                  Pris
                </span>
              </div>
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M16.59 8.59L12 13.17 7.41 8.59 6 10l6 6 6-6z"
                />
              </svg>
              <span
                class="MuiTouchRipple-root"
              />
            </div>
          </nav>
        </div>
        <div
          class="loadingIndicator"
          role="loading"
        >
          <div
            class="makeStyles-root-56"
          >
            <div
              class="MuiCircularProgress-root MuiCircularProgress-colorPrimary MuiCircularProgress-indeterminate"
              role="progressbar"
              style="width: 6rem; height: 6rem;"
            >
              <svg
                class="MuiCircularProgress-svg"
                viewBox="22 22 44 44"
              >
                <circle
                  class="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate"
                  cx="44"
                  cy="44"
                  fill="none"
                  r="20.2"
                  stroke-width="3.6"
                />
              </svg>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
